controllers/user/reset-password.js

```diff
@@ -56,50 +49,72 @@ router
     .get(async (req, res) => {
         let token = req.query.token ? req.query.token : res.locals.token;
         if (!token) {
-            return res.redirect('/user/login');
-        }
-
-        try {
-            await verify(token);
-            res.locals.token = token;
-            renderForm(req, res);
-        } catch (error) {
-            renderFormExpired(req, res);
+            redirect(req, res);
+        } else {
+            try {
+                await verifyToken(token);
+                res.locals.token = token;
+                renderForm(req, res);
+            } catch (error) {
+                renderFormExpired(req, res);
+            }
         }
     })
-    .post(validators.password, async (req, res) => {
-        const changePasswordToken = req.body.token;
+    .post(async (req, res) => {
+        const { token } = req.body;
 
-        if (!changePasswordToken) {
-            res.redirect('/user/login');
+        if (!token) {
+            redirect(req, res);
         } else {
-            // store the token to pass on to other requests (eg. outside the URL)
-            res.locals.token = changePasswordToken;
-            const errors = validationResult(req);
+            try {
```

Comment:
This controller has the biggest structural changes:

- Separates out verification of the token from checking the user
- Decodes the token first and then validates the new password, passes in the username from the token to allow us to check against conditional rules e.g. password must not match username
- Only once we know the format of the data is good do we lookup the user and try to update the password